_stbt/core.py

```diff
@@ -647,29 +671,27 @@ def match(self, image, frame=None, match_parameters=None,
         with numpy_from_sample(frame, readonly=True) as npframe:
             region = Region.intersect(_image_region(npframe), region)
 
-            matched, match_region, first_pass_certainty = _match(
-                _crop(npframe, region), template.image, match_parameters,
-                imglog)
-
-            match_region = Region.from_extents(*match_region).translate(
-                region.x, region.y)
-            result = MatchResult(
-                get_frame_timestamp(frame), matched, match_region,
-                first_pass_certainty, numpy.copy(npframe),
-                (template.name or template.image))
-
-        imglog.set(result=result)
-        _log_match_image_debug(imglog)
-
-        if grabbed_from_live:
-            self._display.draw(result, None)
-
-        if result.match:
-            debug("Match found: %s" % str(result))
-        else:
-            debug("No match found. Closest match: %s" % str(result))
-
-        return result
+            # pylint:disable=undefined-loop-variable
+            try:
+                for (matched, match_region, first_pass_matched,
+                     first_pass_certainty) in _find_matches(
+                        _crop(npframe, region), template.image,
+                        match_parameters, imglog):
+
+                    match_region = Region.from_extents(*match_region) \
+                                         .translate(region.x, region.y)
+                    result = MatchResult(
+                        get_frame_timestamp(frame), matched, match_region,
+                        first_pass_certainty, numpy.copy(npframe),
```

Comment:
We're taking a copy of the frame every time we go round this loop.  Perhaps it would be better to take a single copy at the beginning of the function, mark it read-only and return it for each `MatchResult`.  This would be a (minor) API break for `match()` but I think it could be worth it, particularly given some changes I want to make to how we map the `Gst.Samples` so we don't need to make copies before handing them to users.